### ISSUE TITLE
question - How to enable groupCollapse seperated

### ISSUE THREAD

**Issue by mrjumpy** [free-jqgrid/jqGrid#82]
Hi Oleg.
When `groupCollapse` is true, we can collapse all group.
But, if we want to collapse some group and some just don't collapse.
How to reach that?
Does it difficult to implement?

**Comment by OlegKi** [comment]
Thank you for the interesting question!

The current options of grouping don't allows you to specify different values of `groupCollapse` for different  groups. I can imagine to allow the usage of `groupCollapse` as callback function as an alternative to Boolean value implemented now. I would ask you to set breakpoint on [the line](https://github.com/free-jqgrid/jqGrid/blob/v4.9.0-rc1/js/grid.grouping.js#L358) and to examine which parameters would be good for such callback function. The value `n.idx` defines the level of grouping (typically 0), `i` and `n` are the index in the `p.groupingView.groups` array and the item of `p.groupingView.groups`. The properties `n.value` and `n.displayValue` could be important too.

Look at the options of `formatDisplayField` callback in [the line](https://github.com/free-jqgrid/jqGrid/blob/v4.9.0-rc1/js/grid.grouping.js#L350). Would be the same options of `groupCollapse` callback OK? The callback could return Boolean value which would have the same meaning like static `groupCollapse`. Please examine your case and write me whether such callback with the parameters would solve your problem. After that I could make the corresponding changes in the code of `groupingRender`.

**Comment by mrjumpy** [comment]
I think `callback` implementation will be not that great, 
I thought the developers don't need to `callback function`,  they need `options`.
considering the following scenario:
Imagine that  every time the new options pass into callback function:
```
groupCollapse: function(groupName, level, ...){
    if(groupName == 'baby'){
         //collapase 
     }
}
```

I thought it's too complicated.
Maybe we just provide the developer things like  
`groupCollapse:[{"baby": true, "mom": false}]`.
Because we only know the grouping result when we saw the result right?
So, I think there is no needed the `callback function`
We need to using specific group name to point the options true/false whether collapse.

**Comment by OlegKi** [comment]
The syntax which you suggest `groupCollapse:[{"baby": true, "mom": false}]` is not common enough. What about multilevel grouping (you have something like `ProductName`, `ProductCategory`, `ProductSubcategory`)? Every level of grouping could be collapsed or not. The syntax `{"baby": true, "mom": false}` is not oriented on some specific level or sublevel of grouping.

Moreover the syntax `{"baby": true, "mom": false}` is redundant. What about *another* values as `"baby"` or `"mom"`? Should the groups be collapsed or expanded?

What you mean under `"baby"` or `"mom"`? Do you mean `p.groupingView.groups[i].value`, `p.groupingView.groups[i].displayValue` (formatted value) or some other value?

One have to think about *all* cases of usage the grouping if one introduced some extension? The callback which I suggested in the previous comment could look like
```JavaScript
groupField: ["x0"],
groupCollapse: function(displayValue, value, group, cm, idx, groups) {
    if (group.dataIndex === "x0" && $.inArray(group.value, ["Baby", "none"]) >= 0) {
        return true; // collapse
    }
    return false; // expand
}
```
You could use for example in 
```JavaScript
{
    width: 700,
    pager: true,
    caption: "Basic-pgrid3-bug&enhancement",
    shrinkToFit: false,
    groupingView: {
        collapsedCategories: ["Baby", "none"],
        groupCollapse: function(displayValue, value, group, cm, idx, groups) {
            if (group.dataIndex === "x0" && $.inArray(group.value, groups.collapsedCategories) >= 0) {
               return true; // collapse
            }
            return false; // expand
        }
    }
}
```
In the above example you would be set your custom property in `groupingView`: `collapsedCategories: ["Baby", "none"]` and to use it as `groups.collapsedCategories` in `groupCollapse` callback.

The approach seems me simple and flexible enough. What do you think about?

**Comment by OlegKi** [comment]
I examined the code more exactly and I see that the implementation is not so easy. The problem is the common organization of the data grouping code. The grouping rows will be *first* maked as hidden or not based on "static" Boolean value `groupingView.groupCollapse`. See [the line](https://github.com/free-jqgrid/jqGrid/blob/v4.9.0-rc1/js/grid.base.js#L1877) and [this one](https://github.com/free-jqgrid/jqGrid/blob/v4.9.0-rc1/js/grid.base.js#L1935). *Later* it will be filled the internal information about the group to which the rendering row belong (see the call of `groupingPrepare` [here](https://github.com/free-jqgrid/jqGrid/blob/v4.9.0-rc1/js/grid.base.js#L1943)) and finally one calls [here](https://github.com/free-jqgrid/jqGrid/blob/v4.9.0-rc1/js/grid.base.js#L1955) the method `groupingRender` which insert grouping headers in the array of HTML strings prepared for rendering.

In other words one need make more deep changes in the structure of data grouping code before one will have simple and clear interface to dynamic collapsing of the groups. It's better to make the changes in the next release.

**Comment by OlegKi** [comment]
I implemented recently new feature in data grouping (see [here](https://github.com/free-jqgrid/jqGrid/commit/df91c0dbc5728a30578f19ce37dfbd8eda52230d), [here](https://github.com/free-jqgrid/jqGrid/commit/d35e0f0c19ea8a1f05094dbda2137563582ef677) and [here](https://github.com/free-jqgrid/jqGrid/commit/44cb0098cda6139576169e06af1e5e73398ffdc7)), which allows to use `groupCollapse` as function. It allows to specify, which groups should be collapsed and which expanded based on the values (keys) from the group header.

You can see working of the new feature on [the demo](http://www.ok-soft-gmbh.com/jqGrid/OK/grouping2-groupCollapseAsFunction.htm) and on [this one](http://www.ok-soft-gmbh.com/jqGrid/OK/grouping2-groupCollapseAsFunction1.htm).

The issue is very old and I'm not sure whether you use still free jqGrid or not. Thus I'm closing the issue now. In any way your feedbacks are welcome.
